### ISSUE TITLE
Zip?

### ISSUE THREAD

**Issue by eernstg** [lrhn/dep-pkgspec#13]
Did anybody discuss how to deploy software developed as a package using packages.cfg? I couldn't see any such discusion here.

It used to be possible to zip the `<my-package-name>/build` directory tree and then unzip it on some deployment host; everything would be there, because the `<my-package-name>/build/**/packages` directory/directories would be a full copy, not a single directory with symlinks in it.

With this proposal, doesn't deployment require a tool that is able to understand packages.cfg and create a setup that will work on the deployment host? How could it do that in any other way than (1) copy all the files of packages into a `packages` directory like in the olden days, and, possibly, (2) create one or more build/**/packages.cfg files telling dart to search in that `packages` directory? With that in place, I think a simple zip would work again.

**Comment by zoechi** [comment]
Zip seems not to be an official supported way to deploy a Dart package but sounds interesting.
Currently `pub get` is what initializes the symlinks in the `packages` directory.
But for your internal custom who wants to customize `packages.cfg` this might not be ideal.

**Comment by eernstg** [comment]
`pub get` will set up your local directory for _development_ work, but it
is not obvious that you want the result of `pub get` in your deployment
context (e.g., a public web server). I've heard various people refer to the
approach where deployment happens via `pub build`, `zip .. build`, copy to
deployment host, `unzip ..`, so I assume that it is rather common. This is
also the reason why `build/**/packages' contains a copy of the package
files rather than symlinks.

On Tue, May 12, 2015 at 1:12 PM, Günter Zöchbauer notifications@github.com
wrote:

> Zip seems not to be an official supported way to deploy a Dart package but
> sounds interesting.
> Currently pub get is what initializes the symlinks in the packages
> directory.
> But for your internal custom who wants to customize packages.cfg this
> might not be ideal.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/lrhn/dep-pkgspec/issues/13#issuecomment-101238129.

## 

Erik Ernst  -  Google Danmark ApS
Skt Petri Passage 5, 2 sal, 1165 København K, Denmark
CVR no. 28866984

**Comment by zoechi** [comment]
I see. 
I thought once dart2dart would be it, but back then I was discouraged using it because it's just experimental, but shortly afterwards there was put notable effort into dart2dart. I don't know the current state. For some time I had the impression creating snapshots is the way to go but I'm not sure if this is the preferred or officially supported way.
The Docker images for Manged VM deployment copy the source files into the Docker file and run `pub get` (https://www.dartlang.org/server/google-cloud-platform/app-engine/).

This discussion might not fit to this issue but I guess it should be clarified which deployment story/stories should be supported before discussing how, and what needs to be done to make it work.

**Comment by eernstg** [comment]
I just requested input about this from some of the creators of `pub`.

On Tue, May 12, 2015 at 3:00 PM, Günter Zöchbauer notifications@github.com
wrote:

> I see.
> I thought once dart2dart would be it, but back then I was discouraged
> using it because it's just experimental, but shortly afterwards there was
> put notable effort into dart2dart. I don't know the current state. For some
> time I had the impression creating snapshots is the way to go but I'm not
> sure if this is the preferred or officially supported way.
> The Docker images for Manged VM deployment copy the source files into the
> Docker file and run pub get (
> https://www.dartlang.org/server/google-cloud-platform/app-engine/).
> 
> This discussion might not fit to this issue but I guess it should be
> clarified which deployment story/stories should be supported before
> discussing how, and what needs to be done to make it work.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/lrhn/dep-pkgspec/issues/13#issuecomment-101266933.

## 

Erik Ernst  -  Google Danmark ApS
Skt Petri Passage 5, 2 sal, 1165 København K, Denmark
CVR no. 28866984

**Comment by munificent** [comment]
**TL;DR: We can mostly just keep doing what we're doing and it shouldn't break stuff.**

When it comes to deployment, there's a couple of different kinds of apps you might deploy, and a couple of ways we've envisioned users deploying them.

## Server-side Dart app

For a server-side Dart app where you are running it directly on top of the Dart VM (usually inside some kind of container like Docker or AppEngine), you can deploy it two ways:
1. Run `pub build` and copy the build directory into your container. Currently, that will include all your dependencies since `packages` in the build output contains the actual content.
   
   With packagespec, we'll have to decide what we put in the build output. The easiest solution is to just preserve the current behavior (which will still work).
2. Run `pub get` on the deployment machine (or in the container). Pub was designed around repeatable builds so—as long as you check in your lockfile like a good pub user—doing this will give you an identical setup to what you have on your dev machine.
   
   With packagespec, this will create a packagespec that points to packages in the server's package cache.
   
   This is a pretty typical way of deploying apps in other languages. You just run the package manager right on the server as part of the setup process.

## Client-side Dart app

This is easier. Since the Dart VM is totally out of the loop all discussions of `packages/`, packagespec, etc. go out the window. You run `pub build` and you get a few monolithic JS files that include the code for all of your dependencies.

There's still the question of which _resources_ (non-Dart stuff) to include. For that, we'll likely continue to create a `packages/` directory like we do now. That, along with @lrhn's resource loader stuff should handle non-Dart assets.

## Other stuff

There are other deployment possibilities like mobile apps, embedded devices, etc., but we haven't put much work into those user scenarios yet.

**Comment by zoechi** [comment]
@munificent Not too long ago you mentioned that running `pub build bin` is pretty uncommon http://dartbug.com/18622#c8 . Has this changed? I never saw an announcement that `pub build` can or should be used for server side deployment. Is copying source files and packages all it does for the `bin` directory?

**Comment by munificent** [comment]
Ah, good point. Yeah, that use case is valid but it's not one we've put a lot of effort into. The simplest workaround is to probably just tell the dart2js transformer to not run on the `bin` directory.

**Comment by lrhn** [comment]
This is a discussion for pub.
